consensus/hotstuff/forks/forks2.go

```diff
@@ -0,0 +1,557 @@
+// Notes:
+//   - Method `AddCertifiedBlock(..)` should be used preferably, if a QC certifying
+//     `block` is already known. This is generally the case for the consensus follower.
+//     Method `AddProposal` is intended for active consensus participants, which fully
+//     validate blocks (incl. payload), i.e. QCs are processed as part of validated proposals.
+//
+// Possible error returns:
+//   - model.MissingBlockError if the parent does not exist in the forest (but is above
+//     the pruned view). From the perspective of Forks, this error is benign (no-op).
+//   - model.InvalidBlockError if the block is invalid (see `Forks.EnsureBlockIsValidExtension`
+//     for details). From the perspective of Forks, this error is benign (no-op).
+//   - model.ByzantineThresholdExceededError if conflicting QCs or conflicting finalized
+//     blocks have been detected (violating a foundational consensus guarantees). This
+//     indicates that there are 1/3+ Byzantine nodes (weighted by stake) in the network,
+//     breaking the safety guarantees of HotStuff (or there is a critical bug / data
+//     corruption). Forks cannot recover from this exception.
+//   - All other errors are potential symptoms of bugs or state corruption.
+func (f *Forks2) AddProposal(proposal *model.Block) error {
+	err := f.EnsureBlockIsValidExtension(proposal)
+	if err != nil {
+		return fmt.Errorf("validity check on block %v failed: %w", proposal.BlockID, err)
+	}
+	err = f.UnverifiedAddProposal(proposal)
+	if err != nil {
+		return fmt.Errorf("error storing block %v in Forks: %w", proposal.BlockID, err)
+	}
+	return nil
+}
+
+// UnverifiedAddCertifiedBlock appends the given certified block to the tree of pending
+// blocks and updates the latest finalized block (if applicable). Unless the parent is
+// below the pruning threshold (latest finalized view), we require that the parent is
+// already stored in Forks. Calling this method with previously processed blocks
+// leaves the consensus state invariant (though, it will potentially cause some
+// duplicate processing).
+// We assume that all blocks are fully verified. A valid block must satisfy all
+// consistency requirements; otherwise we have a bug in the compliance layer.
+// Notes:
+//   - UNVALIDATED: expects block to pass `Forks.EnsureBlockIsValidExtension(..)`
+//
+// Error returns:
+//   - model.ByzantineThresholdExceededError if conflicting QCs or conflicting finalized
+//     blocks have been detected (violating a foundational consensus guarantees). This
+//     indicates that there are 1/3+ Byzantine nodes (weighted by stake) in the network,
+//     breaking the safety guarantees of HotStuff (or there is a critical bug / data
+//     corruption). Forks cannot recover from this exception.
+//   - All other errors are potential symptoms of bugs or state corruption.
+func (f *Forks2) UnverifiedAddCertifiedBlock(certifiedBlock *model.CertifiedBlock) error {
```

Comment:
Why this function is public?

Reply:
🤷 No strong reason. We have similar exported methods in the old `Forks` implementation https://github.com/onflow/flow-go/blob/0b531f40e749c12d206236dc196b2e6183f482b2/consensus/hotstuff/forks/forks.go#L109 https://github.com/onflow/flow-go/blob/0b531f40e749c12d206236dc196b2e6183f482b2/consensus/hotstuff/forks/forks.go#L155 Though, we only expose `AddProposal` in the [`Forks` interface](https://github.com/onflow/flow-go/blob/0b531f40e749c12d206236dc196b2e6183f482b2/consensus/hotstuff/forks.go#L43).

When I originally implemented `Forks`, I also exposed the unverified methods to allow higher-level logic to choose which level of safety they desire. Conceptually, we don't need the additional validity check in `Forks`, because the consistency requirements should (in a perfect world 😅) be guaranteed by the compliance layer. 
But from a practical perspective:
* the additional validity check in `Forks` is probably really cheap
* the additional validity check is probably a very sensible sanity check (hardening the implementation)
* we anyway only expose the safe methods in the interface
* we don't use the `Unverified` methods at all (except within forks)

Based on these reasons, I totally agree with you that it would make the most sense to not export these methods.